[JENKINS-44720] Update submodules in parallel

nre-ableton opened the pull request:
## [JENKINS-44720](https://issues.jenkins-ci.org/browse/JENKINS-44720) - support updating submodules in parallel

This change exposes the `threads` property of the `SubmoduleUpdateCommand` (added in https://github.com/jenkinsci/git-client-plugin/pull/348), which allows for git submodules to be updated in parallel. This can greatly improve cloning/checkout times for repositories which have many submodules.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [x] New feature (non-breaking change which adds functionality)

## Further comments

Do I need to also add support for this property in `GitSCMBackwardCompatibility`? I'm not entirely sure as to the purpose of this class, but as far as I can see, it's for older properties which have been deprecated/moved elsewhere. My understanding is that a new property shouldn't need to touch this file, is that correct?

samrocketman commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -42,15 +42,17 @@
     private String reference;
     private boolean parentCredentials;
     private Integer timeout;
+    private Integer threads;
```
I'm a bit confused.  I don't see you using `java.lang.Thread`.  I only see this `Integer` which I assume is meant for a thread count.  I also don't see any usage of thread methods like joining or locking.  Can you explain to me how this is adding parallel submodule updating?

nre-ableton replied:
I agree that it would be nice to match the number of cores, but I'm not sure this is very practical. Even if there is an API call to get the number of cores (I have no idea if there is), it would only make sense if the build is executed on the Jenkins master. If the build is run on a remote node, then the user would obviously want to set this value based on the core count there.

I think that 1 is still a sensible default, though I would agree that 8 is a sensible maximum value, since anything more than that is likely to cause problems elsewhere. I'll add a fixup to add this.

darxriggs replied:
I also think that 1 should be the default.
The upper bound shouldn't be limited and completely be up to the user.
@nre-ableton maybe you misunderstood @samrocketman when he was talking about the maximum value. He talked about limiting the value when the smart default is not 1 but detected in a smart way.

nre-ableton replied:
@darxriggs Yes, sorry, I think I did misunderstand. Thanks for the clarification. I'll remove my fixup.

samrocketman replied:
Yes, I meant we did not want to auto-detect abusing a user's system so limiting it without setting too high of a default.  I think we're all on the same page, though :smile:

darxriggs commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -42,15 +42,17 @@
 
     @DataBoundConstructor
-    public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference,Integer timeout, boolean parentCredentials) {
+    public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference, Integer timeout, boolean parentCredentials, Integer threads) {
```
The constructor should only be extended for mandatory parameters.
For this optional parameter a setter with `@DataBoundSetter` should be used.

nre-ableton replied:
@darxriggs Ok, I've made this change and amended it to the previous commit. The tests still pass, so I hope that I did it right. 😅

samrocketman replied:
I tend to update primary constructors like the following:

```java
@DataBoundConstructor
public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference, Integer timeout, boolean parentCredentials, Integer threads) {
    // cool code goes here
}

//but we still need to migrate older configurations so...
@Deprecated
public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference,Integer timeout, boolean parentCredentials) {
    this(disableSubmodules, recursiveSubmodules, trackingSubmodules, reference, timeout, parentCredentials, 1);
    // don't forget the 1 is to set migrating configurations to 1 thread by default.
}
```

samrocketman commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -42,18 +42,20 @@
 
     @DataBoundConstructor
-    public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference,Integer timeout, boolean parentCredentials) {
+    public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference, Integer timeout, boolean parentCredentials) {
```
I tend to update primary constructors like the following:

```java
@DataBoundConstructor
public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference, Integer timeout, boolean parentCredentials, Integer threads) {
    // cool code goes here
}

//but we still need to migrate older configurations so...
@Deprecated
public SubmoduleOption(boolean disableSubmodules, boolean recursiveSubmodules, boolean trackingSubmodules, String reference,Integer timeout, boolean parentCredentials) {
    this(disableSubmodules, recursiveSubmodules, trackingSubmodules, reference, timeout, parentCredentials, 1);
    // don't forget the 1 is to set migrating configurations to 1 thread by default.
}
```

darxriggs replied:
Constructors should only be extended for mandatory parameters - see https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#constructor-vs-setters. 
This also limits the need to deprecate old constructor versions.
Unfortunately this rule hasn't be followed in the past for many plugins.

samrocketman replied:
Interesting, thanks for the documentation link.  I was not aware before but am now.

darxriggs commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -45,15 +45,17 @@
         this.parentCredentials = parentCredentials;
         this.reference = reference;
         this.timeout = timeout;
+        this.threads = 1;
```
There is already a default value set in line 48 so this is not required here. It duplicates code and increases the potential for inconsistencies.

Please remove this line.

darxriggs commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -45,15 +45,17 @@
     /** Use --depth flag on submodule update command - requires git>=1.8.4 */
     private boolean shallow;
     private Integer depth;
+    private Integer threads = 1;
```
I wouldn't use a default value here. Due to `setThreads(Integer threads)` accepting `null` and negative numbers, a proper value check is anyway required before handing it to `SubmoduleUpdateCommand::threads`.

Please remove the default value assignment.

darxriggs commented on src/main/java/hudson/plugins/git/extensions/impl/SubmoduleOption.java:
```diff
@@ -126,6 +137,7 @@ public void onCheckoutCompleted(GitSCM scm, Run<?, ?> build, GitClient git, Task
                         .parentCredentials(parentCredentials)
                         .ref(build.getEnvironment(listener).expand(reference))
                         .timeout(timeout)
+                        .threads(threads)
```
Could you validate the set `threads` value (similar to the one for `depth`)?
I suggtest to replace line 140 with something like this between line 146 and 147:
```
int usedThreads = threads == null || threads < 1 ? 1 : threads;
cmd.threads(usedThreads);
```
The `null` check will be required anyway because I suggest to change the method signature for `SubmoduleUpdateCommand::threads` from taking an `Integer` to taking an `int` because `null` values are not required in there.

nre-ableton commented:
@darxriggs Thanks for the review! I've applied your suggestions, rebased, and added some links to the PRs in `git-client-plugin`. Is there anything else that needs to be done?

darxriggs commented:
Thanks for your patience. This PR looks fine now. We'll leave it open until the changes in jenkinsci/git-client-plugin#352 are merged.

darxriggs commented:
Hi @MarkEWaite, I just wanted to remind you that even if this pull request is not merged and threads not visible to the user, the changes of jenkinsci/git-client-plugin#348 are already in effect. If jenkinsci/git-client-plugin#352 (and potentially additional required changes) doesn't make it into the v4.0.0 release, the first pull request should be reverted.

kshultzCB commented:
@darxriggs, your review status is still showing "Requested changes." Is that still the case?

darxriggs commented:
@kshultzCB, I requested changes in this pull request - all of them have been applied - see https://github.com/jenkinsci/git-plugin/pull/612#issuecomment-423056229.

But
a) I don't have the rights to actually approve this pull request
b) this pull request should not be merged until the issues in the git-client-plugin's pull request jenkinsci/git-client-plugin#352 are resolved

nre-ableton commented:
@MarkEWaite thanks for all your work on this PR! I'm so glad to finally see it get merged. 🎉
